### ISSUE TITLE
Rethink early finish bonus options

### ISSUE THREAD

**Issue by Slayer95** [wesnoth/wesnoth#7590]
### Describe the desired feature

Currently, ``[endlevel] bonus`` accepts ``boolean|number`` as parameters.
- ``yes``. Result equivalent to "instantly capture all villages, delete all your non-loyal units, and wait til the last turn & win." Except that no units are sacrificed in the process, so the player will _NOT_ wait.
- ``number`` is a factor that multiplies the benefit from ``yes``. It's hard to figure out what an appropriate value for it should be.
- ``no``. In practice, gameplay is roughly equivalent to "ensure you have a positive income, capture all villages, and wait til the last turn, then win." This (i.e. _waiting_) is what many players do when running a user-made-campaign with early finish bonus off.

IMO, waiting at the end of scenarios is very anticlimatic for any game. Therefore, it should be discouraged. However, the alternative of ``yes`` will often provide a huge gold boost for the player. Furthermore, the fact that maintenance cost isn't subtracted from the bonus is not intuitive.

So I think that ``[endlevel]`` should get a ``bonus_per_turn`` attribute, with the following options:
- ``max_gold`` (or ``max_revenue``). This is the current behavior.
- ``max_income``: Result equivalent to "1) Instantly capture all villages 2) If you have a positive net income, then wait til the last turn & win, else, win immediately."
- ``income``: Result equivalent to "If you have a positive net income, then wait til the last turn & win; else, win immediately."

With these options available, a UMC author that doesn't want a disruptive setting of early finish bonus MUST prefer ``income`` over ``no``. Then there would be no reason for players to stall out scenarios in an effort to farm gold. Also, ``max_income`` MAY be preferred over ``max_revenue`` when the scale of the scenario implies there will be lots of villages and units.

**Comment by gfgtdf** [comment]
Somewhat related, i think the whole carryover mechanic, including the dialog showing at the end should be optional. IMO ideally the carryover_show_gold function would be implemented in lua (in a "scenario_end" event) to make it easy for umc devs to implement their won early bonuses / carryover mechanics. I'm thinking for example about wc2 here that splits carryover gold between players but also about other campaigns that don't want any gold carryover at all.

**Comment by Pentarctagon** [comment]
@nemaara @knyghtmare @Konrad22 any thoughts on this idea?

**Comment by nemaara** [comment]
There are a lot of ways to solve the problem with campaign design but I agree that more options is good for designers.

**Comment by Slayer95** [comment]
> Slayer95's argument is correct as many of the outdated mainline campaigns allow this kind of "trolling" gameplay. You camp at enemy leader keep castle and wait for enemy level 2 recruit and then farm the XP while having 100% village control and huge flow of positive income. An example is the Princess of Wesnoth in HttT.

Indeed. Regarding XP, there has been some (failed?) experiments such as [TRoW with a points system](https://forums.wesnoth.org/viewtopic.php?f=8&t=51087). However, gold can be more simply scoped, so I'd like to fix at least the gold aspect of early finish. 

> Whether the existing the system should be reworked or not is contingent upon how quickly, and how effectively rebalance of the several mainline campaigns can be done.

It also wasn't originally my focus to rebalance campaigns that already provide early finish bonuses. However, in cases where <i>"the scale of the scenario implies there will be lots of villages and units"</i>, then that would certainly be an option. Also WFL ``bonus_per_turn=(income + 5)`` would be much simpler to reason about for content creators (5 gold per turn additional to what you'd get by waiting til end of turns).

To clarify, I'd like people to stop using ``bonus=no``, unless they implement custom carryover mechanics as mentioned by @gfgtdf  . As for the current campaigns, I believe balancing bonus through baseline ``income + N`` is superior, but ultimately using it will depend on willingness to do such a change.
